src/libertem/udf/stddev.py

```diff
@@ -0,0 +1,171 @@
+# import functools
+import collections
+
+import numba
+import numpy as np
+
+from libertem.common.buffers import BufferWrapper
+
+# from libertem.udf import ResultBuffer, map_frames
+
+VariancePart = collections.namedtuple('VariancePart', ['sum_var', 'sum_im', 'N'])
+
+
+def my_buffer_batch():
+    return {
+        'batch' : BufferWrapper(
+            kind = 'sig', extra_shape = (3,), dtype = 'float32'
+            )
+    }
+
+def my_frame_fn_batch(frame, batch):
+
+    if batch[:, :, 2][0][0] == 0:
+        batch[:, :, 0] = 0
+
+    else:
+        
+        p0 = VariancePart(sum_var = batch[:, :, 0], sum_im = batch[:, :, 1], N = batch[:, :, 2][0][0])
+        p1 = VariancePart(sum_var = 0, sum_im = frame, N = 1)
+        compute_merge = merge(p0, p1)
+
+        sum_var, sum_im, N = compute_merge.sum_var, compute_merge.sum_im, compute_merge.N
+
+        batch[:, :, 0] = sum_var
+
+    batch[:, :, 1] += frame
+    batch[:, :, 2] += 1
+
+def stddev_merge(dest, src):
+
+    p0 = VariancePart(sum_var = dest['batch'][:, :, 0], sum_im = dest['batch'][:, :, 1], N = dest['batch'][:, :, 2][0][0])
+    p1 = VariancePart(sum_var = src['batch'][:, :, 0], sum_im = src['batch'][:, :, 1], N = src['batch'][:, :, 2][0][0])
+    compute_merge = merge(p0, p1)
+
+    sum_var, sum_im, N = compute_merge.sum_var, compute_merge.sum_im, compute_merge.N
+    dest['batch'][:, :, 0] = sum_var
+    dest['batch'][:, :, 1] = sum_im
+    dest['batch'][:, :, 2] = N
+
+def minibatch(partition):
```

Comment:
For now, I've removed this function and "part" function. I think the main problem with current implementation is that run_udf gives us partitions of the frames but we can't access a partition at its entirety; for instance, we need to use the passed function (my_frame_fn_batch) on run_udf to perform "merge" for individual frame at the partition level and then perform another "merge" with other partitions. If we can access a partition at a time (i.e., view the whole frames within the partition at once), then this would be much cleaner process in terms of code and perhaps in terms of speed as well.

Reply:
We thought about providing a per-partition interface, too, and we may add one in the future. We want to start smaller, though, and with https://github.com/LiberTEM/LiberTEM/issues/310 we plan to introduce an interface that gives you a tile (think small list of frames). This should then allow to implement a minibatch-like function, and improve both readability and performance. I'd be curious by how much; let's benchmark this once #310 is done.

I also propose a different by-frame interface in [a comment on #310](https://github.com/LiberTEM/LiberTEM/issues/310#issuecomment-478175692); do you think this would help with writing cleaner code? I don't think it would improve performance, though.

The reason we don't give access to the whole partition as a big buffer is, that it is easy to write inefficient code that way, which does multiple passes over a partition, thus easily being limited by memory bandwidth. By only operating on smaller tiles or single frames we make sure the current working set at least fits into the L3 cache of the CPU.